android-o/AutofillFramework/AutofillService/Data/Adapter/DatasetAdapter.cs

```diff
@@ -0,0 +1,224 @@
+				fieldTypesByAutofillHint = fieldTypesByAutofillHint,
+				filledAutofillFieldsByTypeName = filledAutofillFieldsByTypeName,
+				datasetAdapter = this
+
+			});
+
+			return setValueAtLeastOnce.Value;
+		}
+
+		private bool BindDatasetToFocusedNode(FilledAutofillField field, FieldType fieldType, Dataset.Builder builder)
+		{
+			MutableBoolean setValueAtLeastOnce = new MutableBoolean(false);
+			mClientParser.Parse(new BindDatasetToFocusedNodeProcessor()
+			{
+				field = field,
+				builder = builder,
+				datasetAdapter = this,
+				setValueAtLeastOnce = setValueAtLeastOnce
+			});
+			return setValueAtLeastOnce.Value;
+		}
+
+		public class BindDatasetNodeProcessor : ClientParser.INodeProcessor
+		{
+			public Dataset.Builder datasetBuilder;
+			public MutableBoolean setValueAtLeastOnce;
+			public Dictionary<string, FieldTypeWithHeuristics> fieldTypesByAutofillHint;
+			public Dictionary<string, FilledAutofillField> filledAutofillFieldsByTypeName;
+
+			public DatasetAdapter datasetAdapter;
+
+			public void ProcessNode(AssistStructure.ViewNode node)
+			{
+				datasetAdapter.ParseAutofillFields(node, fieldTypesByAutofillHint, filledAutofillFieldsByTypeName,
+					datasetBuilder, setValueAtLeastOnce);
+			}
+		}
+
+		public class BindDatasetToFocusedNodeProcessor : ClientParser.INodeProcessor
+		{
+			public FilledAutofillField field;
+			public Dataset.Builder builder;
+			public DatasetAdapter datasetAdapter;
+			public MutableBoolean setValueAtLeastOnce;
+
+			public void ProcessNode(AssistStructure.ViewNode node)
+			{
+				if (node.IsFocused && node.AutofillId != null)
+				{
+					datasetAdapter.BindValueToNode(node, field, builder, setValueAtLeastOnce);
+				}
+			}
+		}
+
+		private void ParseAutofillFields(AssistStructure.ViewNode viewNode,
+			Dictionary<string, FieldTypeWithHeuristics> fieldTypesByAutofillHint,
+			Dictionary<string, FilledAutofillField> filledAutofillFieldsByTypeName,
+			Dataset.Builder builder, MutableBoolean setValueAtLeastOnce)
+		{
+			var rawHints = viewNode.GetAutofillHints();
+			if (rawHints == null || rawHints.Length == 0)
+			{
+				Util.Logv("No af hints at ViewNode - %s", viewNode.IdEntry);
+				return;
+			}
+			string fieldTypeName = AutofillHints.GetFieldTypeNameFromAutofillHints(
+				fieldTypesByAutofillHint, rawHints.ToList());
+			if (fieldTypeName == null)
+			{
+				return;
+			}
+			FilledAutofillField field = filledAutofillFieldsByTypeName[fieldTypeName];
+			if (field == null)
+			{
+				return;
+			}
+			BindValueToNode(viewNode, field, builder, setValueAtLeastOnce);
+		}
+
+		void BindValueToNode(AssistStructure.ViewNode viewNode,
+			FilledAutofillField field, Dataset.Builder builder,
+			MutableBoolean setValueAtLeastOnce)
+		{
+			AutofillId autofillId = viewNode.AutofillId;
+			if (autofillId == null)
+			{
+				Util.Logw("Autofill ID null for %s", viewNode.ToString());
+				return;
+			}
+			int autofillType = (int)viewNode.AutofillType;
+			switch (autofillType)
```

Comment:
Change this to `switch (viewNode.AutofillType)` and you can remove all the casting here.